From 923c16d278b40d4d9105e58dfbbe4c1b8895ac55 Mon Sep 17 00:00:00 2001
From: hubin <hubin73@huawei.com>
Date: Tue, 6 May 2025 20:22:07 +0800
Subject: [PATCH 929/960] downstream: acpi: provides acpi power notifier chain

To support BMC poweroff notification under UB architecture, we need to
provide notifier chain for driver to register and block poweroff until
notifier chain returns success. And to avoid userspace service getting
shutdown before blocking poweroff procedure, notifier chain should be
called before input_report_key, which is earlier than netlink event.

Signed-off-by: hubin <hubin73@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 config.aarch64        |  1 +
 drivers/acpi/Kconfig  |  7 +++++++
 drivers/acpi/button.c | 23 +++++++++++++++++++++++
 include/acpi/button.h | 16 ++++++++++++++++
 4 files changed, 47 insertions(+)

diff --git a/config.aarch64 b/config.aarch64
index 95ad6baf0fd8..8ee2079af487 100644
--- a/config.aarch64
+++ b/config.aarch64
@@ -689,6 +689,7 @@ CONFIG_ACPI_AGDI=y
 CONFIG_ACPI_APMT=y
 CONFIG_ACPI_PPTT=y
 CONFIG_ACPI_PCC=y
+CONFIG_ACPI_POWER_NOTIFIER_CHAIN=y
 # CONFIG_ACPI_FFH is not set
 # CONFIG_PMIC_OPREGION is not set
 CONFIG_ACPI_PRMT=y
diff --git a/drivers/acpi/Kconfig b/drivers/acpi/Kconfig
index e66874d4c3e3..e74d1873bc3c 100644
--- a/drivers/acpi/Kconfig
+++ b/drivers/acpi/Kconfig
@@ -181,6 +181,13 @@ config ACPI_BUTTON
 	  To compile this driver as a module, choose M here:
 	  the module will be called button.
 
+config ACPI_POWER_NOTIFIER_CHAIN
+	bool "enable acpi power notifier chain"
+	depends on ACPI_BUTTON && ARM64
+	default n
+	help
+	  Say Y here to enable acpi power notifier chain.
+
 config ACPI_TINY_POWER_BUTTON
 	tristate "Tiny Power Button Driver"
 	depends on !ACPI_BUTTON
diff --git a/drivers/acpi/button.c b/drivers/acpi/button.c
index c760e38df981..2394ad368d6b 100644
--- a/drivers/acpi/button.c
+++ b/drivers/acpi/button.c
@@ -436,6 +436,22 @@ static void acpi_lid_notify(acpi_handle handle, u32 event, void *data)
 	acpi_lid_update_state(device, true);
 }
 
+#if IS_ENABLED(CONFIG_ACPI_POWER_NOTIFIER_CHAIN)
+static BLOCKING_NOTIFIER_HEAD(acpi_power_chain_head);
+
+int register_acpi_power_notifier(struct notifier_block *nb)
+{
+	return blocking_notifier_chain_register(&acpi_power_chain_head, nb);
+}
+EXPORT_SYMBOL(register_acpi_power_notifier);
+
+int unregister_acpi_power_notifier(struct notifier_block *nb)
+{
+	return blocking_notifier_chain_unregister(&acpi_power_chain_head, nb);
+}
+EXPORT_SYMBOL(unregister_acpi_power_notifier);
+#endif
+
 static void acpi_button_notify(acpi_handle handle, u32 event, void *data)
 {
 	struct acpi_device *device = data;
@@ -449,6 +465,13 @@ static void acpi_button_notify(acpi_handle handle, u32 event, void *data)
 		return;
 	}
 
+#if IS_ENABLED(CONFIG_ACPI_POWER_NOTIFIER_CHAIN)
+	if (blocking_notifier_call_chain(&acpi_power_chain_head, 0, 0) == NOTIFY_BAD) {
+		pr_info("acpi power notifier chain: receive bad result, stop poweroff\n");
+		return;
+	}
+#endif
+
 	acpi_pm_wakeup_event(&device->dev);
 
 	button = acpi_driver_data(device);
diff --git a/include/acpi/button.h b/include/acpi/button.h
index af2fce5d2ee3..789ed88d0d57 100644
--- a/include/acpi/button.h
+++ b/include/acpi/button.h
@@ -2,6 +2,8 @@
 #ifndef ACPI_BUTTON_H
 #define ACPI_BUTTON_H
 
+#include <linux/notifier.h>
+
 #define ACPI_BUTTON_HID_POWER	"PNP0C0C"
 #define ACPI_BUTTON_HID_LID	"PNP0C0D"
 #define ACPI_BUTTON_HID_SLEEP	"PNP0C0E"
@@ -15,4 +17,18 @@ static inline int acpi_lid_open(void)
 }
 #endif /* IS_ENABLED(CONFIG_ACPI_BUTTON) */
 
+#if IS_ENABLED(CONFIG_ACPI_POWER_NOTIFIER_CHAIN)
+int register_acpi_power_notifier(struct notifier_block *nb);
+int unregister_acpi_power_notifier(struct notifier_block *nb);
+#else
+static inline int register_acpi_power_notifier(struct notifier_block *nb)
+{
+	return -EINVAL;
+};
+static inline int unregister_acpi_power_notifier(struct notifier_block *nb)
+{
+	return -EINVAL;
+};
+#endif /* IS_ENABLED(CONFIG_ACPI_POWER_NOTIFIER_CHAIN) */
+
 #endif /* ACPI_BUTTON_H */
-- 
2.43.0

